### PR TITLE
allow custom colorscale in KepplerMapper.visualize

### PR DESCRIPTION
**Pull request by galtay**
Allows the user to pass in a custom colorscale. 

Sorry for the whitespace change noise. I don't have strong opinions about this stuff, but my editor has its settings. I can submit a different PR w/o these whitespace changes if it helps. I'll put comments on the changes relevant to colorscale.

In some places I used `colorscale=colorscale_default` in function signatures instead of None. If the mutable default is a problem I can either change back to None and put the `if colorscale is None` logic in each function OR write a function in visuals called `_return_default_colorscale` and call that function as the kwarg default value ... so  `colorscale=_return_default_colorscale()` instead of `colorscale=None`.

### CODE REVIEW

File: kmapper/kmapper.py [hunk 1]
```diff
@@ -615,6 +615,7 @@ def visualize(
         self,
         graph,
         color_function=None,
+        colorscale=colorscale_default,
```
Comment (galtay) [inline kmapper/kmapper.py:618]: default to the colorscale defined in visuals. this is where we could potentially do `colorscale=_return_default_colorscale()` if we didn't want a mutable default value for a kwarg

File: kmapper/kmapper.py [hunk 2]
```diff
@@ -637,11 +638,14 @@ def visualize(
         color_function : list or 1d array
             A 1d vector with length equal to number of data points used to build Mapper. Each value should correspond to a value for each data point and color of node is computed as the average value for members in a node.
 
+        colorscale : list
+            Specify the colorscale to use. See visuals.colorscale_default.
+
```
Comment (galtay) [inline kmapper/kmapper.py:643]: point the reader to the default colorscale as an example

File: kmapper/kmapper.py [hunk 3]
```diff
@@ -750,6 +754,7 @@ def visualize(
         mapper_data = format_mapper_data(
             graph,
             color_function,
+            colorscale,
```
Comment (galtay) [inline kmapper/kmapper.py:757]: pass colorscale to `format_mapper_data` to get colorscale into cluster details view

File: kmapper/kmapper.py [hunk 4]
```diff
@@ -759,8 +764,6 @@ def visualize(
             nbins,
         )
 
-        colorscale = colorscale_default
-
```
Comment (galtay) [inline kmapper/kmapper.py:763]: remove default colorscale asignment

File: kmapper/visuals.py [hunk 5]
```diff
@@ -165,7 +165,7 @@ def format_meta(graph, custom_meta=None, color_function_name=None):
 
 
 def format_mapper_data(
-    graph, color_function, X, X_names, lens, lens_names, custom_tooltips, env, nbins=10
+    graph, color_function, colorscale, X, X_names, lens, lens_names, custom_tooltips, env, nbins=10
```
Comment (galtay) [inline kmapper/visuals.py:168]: add arg to `format_mapper_data` to get colorscale into cluster details view

File: kmapper/visuals.py [hunk 6]
```diff
@@ -184,6 +184,7 @@ def format_mapper_data(
             lens,
             lens_names,
             color_function,
+            colorscale,
```
Comment (galtay) [inline kmapper/visuals.py:187]: for cluster details view

File: kmapper/visuals.py [hunk 7]
```diff
@@ -210,13 +211,10 @@ def format_mapper_data(
     return json_dict
 
 
-def build_histogram(data, colorscale=None, nbins=10):
+def build_histogram(data, colorscale=colorscale_default, nbins=10):
```
Comment (galtay) [inline kmapper/visuals.py:214]: can leave like this, or go back to `None` or use `_return_default_colorscale`

File: kmapper/visuals.py [hunk 8]
```diff
@@ -377,6 +375,7 @@ def _format_tooltip(
     lens,
     lens_names,
     color_function,
+    colorscale,
```
Comment (galtay) [inline kmapper/visuals.py:378]: for cluster detail view

**Comment by galtay** [comment]
looks like tests are failing b/c lack of default args, I can thread the default through or explicitly pass the value in the test

File: kmapper/kmapper.py (hunk 1 above)
Comment (sauln) [reply kmapper/kmapper.py:618]: I would definitely prefer the function return :D. The mutable default values was a bit of a sticking point during our JOSS submission.

File: kmapper/visuals.py (hunk 7 above)
Comment (sauln) [reply kmapper/visuals.py:214]: The `None` pattern I think will render better in the documentation.

**Comment by galtay** [comment]
great, thanks for taking a look. I'll switch the defaults to None!

File: kmapper/kmapper.py (hunk 2 above)
Comment (sauln) [reply kmapper/kmapper.py:643]: It'd be nice to specify the full path to get to that function so it's a bit clearer. Maybe add a markdown style link like
`[visuals.colorscale_default]( https://github.com/scikit-tda/kepler-mapper/blob/51dde7d0469feeb7d76ba1932bae4ba7cc71e970/kmapper/visuals.py#L10)`?  I'm not sure if that's the right method, but I could imagine newcomers getting hung up on how to find the `colorscale_default`.

**Comment by sauln** [comment]
Looks great, thanks for changing the default parameter.  You think it's ready to go out?  (I'm not worried about a small diff in code coverage)

**Comment by galtay** [comment]
:+1: No problem, thanks for reviewing! I think its good to go. I've run it locally with default colorscale and one I derived from matplotlib (the dark purple nodes were getting buried in the dark background). FWIW this is my hacky code to create new colorscales from matplotlib colormaps  

```python
#cmap = plt.cm.jet
#cmap = plt.cm.rainbow
cmap = plt.cm.gist_rainbow

ii_off = 0 # 20
ff_off = 60
ii = 0 + ii_off
ff = cmap.N - ff_off
sk = (cmap.N - ii_off - ff_off) // 11
cmap_list = [
    cmap(el) for el in np.arange(cmap.N)[ii:ff:sk]
]
rgb_strings = [
    "rgb({}, {}, {})".format(
        int(255 * el[0]), int(255 * el[1]), int(255 * el[2])
    ) for el in cmap_list
]
colorscale = list(zip(np.arange(11)/10, rgb_strings))
```

**Comment by sauln** [comment]
ohhh! How hacky is it?  If it works, do you want to add it to the code with a bit of docks so others can generate their own as well?

**Comment by galtay** [comment]
well, it works :) ... I could add a function to visuals.py ... something like `colorscale_from_matplotlib_cmap`.  its hard coded to have 10 bins b/c I didn't quite understand if that was required but that's what colorscale_default has.

**Comment by galtay** [comment]
the `ii_off` and `ff_off` let you clip off the beginning or end of the cmap if you want to remove dark or light colors (the example above removes the dark blue/purple from gist_rainbow https://matplotlib.org/tutorials/colors/colormaps.html

File: kmapper/visuals.py [hunk 9]
```diff
@@ -58,6 +58,28 @@
 ]
 
 
+def colorscale_from_matplotlib_cmap(cmap, ii_off=0, ff_off=0, nbins=10):
+    """Create a colorscale from a matplotlib colormap.
+    see https://matplotlib.org/tutorials/colors/colormaps.html
+
+    Example:
+      import matplotlib.pyplot as plt
+      colorscale = colorscale_from_matplotlib_cmap(plt.cm.cool)
```
Comment (sauln) [inline kmapper/visuals.py:67]: Could you add your explanation of what `ii_off` and `ff_off` are within a comment here. You can format the params like the others: https://github.com/scikit-tda/kepler-mapper/blob/51dde7d0469feeb7d76ba1932bae4ba7cc71e970/kmapper/kmapper.py#L87

**Comment by sauln** [comment]
Awesome! that looks great.  Just a bit of formatting for the docstring and I think it's good to go.  

Have you tried with more than 10 bins?  I don't think it's required, but not sure if there's some other side effects I can't remember.

**Comment by galtay** [comment]
thanks, will update the docstring.  I just tried with 20 bins and don't see any change in the html histograms (member distribution and node distribution). maybe its interpolated back to 10 somewhere in the code?

**Comment by galtay** [comment]
Added a few checks to the colorscale builder function. Probably still some wacky corner cases, but should be fine with most matplotlib colormaps

**Comment by sauln** [comment]
Sweet! Thanks for your work on this :D

**Pull request merged**